[SwiftUI] Support download functionality

rr-codes opened the pull request:
#### b577152034d7c8f68bf50523cc8d6f4481a80f66
<pre>
[SwiftUI] Support download functionality
<a href="https://bugs.webkit.org/show_bug.cgi?id=285225">https://bugs.webkit.org/show_bug.cgi?id=285225</a>
<a href="https://rdar.apple.com/141109043">rdar://141109043</a>

Reviewed by Aditya Keerthi.

In WebKit, there are several ways of initiating a download for a particular resource:

* Using the WebKit API directly to create a download using a specific `URLRequest`.

* Using the WebKit API directly to create a download using the data from a previously-cancelled download.

* A download can be created via web content before a request is sent by indicating a navigation action policy of `.download`.
This is typically used to download from a download attribute on an `&lt;a&gt;` tag.

* A download can be created via web content in the response after the HTTP headers have been received by indicating a navigation
response policy of `.download`. This is typically used to download a file with a &apos;Content-Disposition: attachment&apos; HTTP header.

* A download can be created using one of the &quot;Download linked file&quot; default context menu items.

======

The new interface for creating and managing downloadable resources is described as follows:

* A download may be created or resumed by an API client using the respective functions on `WebPage`; these return an identifier
for the new download.

* In implementations of the navigation action and navigation response policy decision functions, clients can return a value of
`.download` to indicate the web content should be downloaded.

* A new protocol `DownloadCoordinator` is introduced.

* The `destination(forDownload:)` function of `DownloadCoordinator` is required to be implemented by clients so that WebKit can know
where to place the download. If this is not implemented, downloads are immediately cancelled. Downloads are also cancelled and an
exception raised if the file URL can not be created or already exists.

* The remaining functions of `DownloadCoordinator` are used to implement various policy decisions.

* WebPage now has an observable `downloads` property which is an async sequence of download events. Each download event has a
pecific kind to represent the download&apos;s progress, and an associated download value.

======

SwiftBrowser is updated to excercise this new functionality:

* Various types are refactored for better organization

* The app&apos;s sandbox is expanded to allow for access to the user&apos;s Downloads directory.

* Downloads now work as expected and download into the Downloads directory.

* A new Downloads window is added to view current downloads, with functionality to cancel ongoing downloads, and show finished downloads
in Finder.

=====

Misc:

* Remove a debug assertion that gets triggered when activating a context menu item that begins a download. While the comment above it
still holds true, the codepath will still be called regardless even though it doesn’t actually do anything, which is expected.

* Source/WebKit/UIProcess/API/Swift/DownloadCoordinator.swift: Added.
(DownloadCoordinator.destination(forDownload:response:suggestedFilename:)):
(DownloadCoordinator.authenticationChallengeDisposition(forDownload:challenge:URLCredential:)):
(DownloadCoordinator.httpRedirectionPolicy(forDownload:response:newRequest:)):
(DownloadCoordinator.placeholderPolicy(forDownload:)):
* Source/WebKit/UIProcess/API/Swift/WKDownloadDelegateAdapter.swift: Added.
(WKDownloadDelegateAdapter.owner):
(WKDownloadDelegateAdapter.downloadDidFinish(_:)):
(WKDownloadDelegateAdapter.download(_:didReceivePlaceholderURL:)):
(WKDownloadDelegateAdapter.download(_:didFailWithError:resumeData:)):
(WKDownloadDelegateAdapter.download(_:didReceiveFinalURL:)):
(WKDownloadDelegateAdapter.download(_:decideDestinationUsing:suggestedFilename:)):
(WKDownloadDelegateAdapter.download(_:willPerformHTTPRedirection:newRequest:)):
(WKDownloadDelegateAdapter.download(_:respondTo:URLCredential:)):
(WKDownloadDelegateAdapter.placeholderPolicy(forDownload:URL:)):
* Source/WebKit/UIProcess/API/Swift/WKNavigationDelegateAdapter.swift:
(WKNavigationDelegateAdapter.yieldDownloadProgress(_:download:)):
(WKNavigationDelegateAdapter.webView(_:navigationAction:didBecome:)):
(WKNavigationDelegateAdapter.webView(_:navigationResponse:didBecome:)):
* Source/WebKit/UIProcess/API/Swift/WebPage+Download.swift: Added.
(originalRequest):
(isUserInitiated):
(originatingFrame):
(progress):
(id):
(cancel):
(Downloads.makeAsyncIterator):
(Iterator.source):
(Iterator.next):
* Source/WebKit/UIProcess/API/Swift/WebPage+FrameInfo.swift:
(FrameInfo.wrapped):
* Source/WebKit/UIProcess/API/Swift/WebPage.swift:
(startDownload(using:)):
(resumeDownload(fromResumeData:)):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/SwiftBrowser/Resources/SwiftBrowser.entitlements:
* Tools/SwiftBrowser/Source/DownloadsView.swift: Added.
(DownloadsView.body):
* Tools/SwiftBrowser/Source/Extensions/URLResponse+Extras.swift: Copied from Source/WebKit/UIProcess/API/Swift/WebPage+FrameInfo.swift.
(URLResponse.hasAttachment):
* Tools/SwiftBrowser/Source/SwiftBrowser.swift:
(SwiftBrowserApp.body):
* Tools/SwiftBrowser/Source/ViewModel/AppStorageKeys.swift: Copied from Tools/SwiftBrowser/Source/AppStorageKeys.swift.
* Tools/SwiftBrowser/Source/ViewModel/BrowserViewModel.swift: Renamed from Tools/SwiftBrowser/Source/BrowserViewModel.swift.
(PDF.transferRepresentation):
(BrowserViewModel.decideSensorAuthorization(_:frame:origin:)):
(BrowserViewModel.displayedURL):
(BrowserViewModel.exportedPDF):
(BrowserViewModel.currentDialog):
(BrowserViewModel.currentFilePicker):
(BrowserViewModel.openURL(_:)):
(BrowserViewModel.didReceiveNavigationEvent(_:)):
(BrowserViewModel.navigateToSubmittedURL):
(BrowserViewModel.exportAsPDF):
(BrowserViewModel.didExportPDF(_:any:)):
(BrowserViewModel.didImportFiles(_:any:)):
(BrowserViewModel.setCameraCaptureState(_:)):
(BrowserViewModel.setMicrophoneCaptureState(_:)):
* Tools/SwiftBrowser/Source/ViewModel/DialogPresenter.swift: Added.
(hash(into:)):
(owner):
(handleJavaScriptAlert(_:initiatedBy:)):
(handleJavaScriptConfirm(_:initiatedBy:)):
(handleJavaScriptPrompt(_:defaultText:initiatedBy:)):
(handleFileInputPrompt(_:initiatedBy:)):
* Tools/SwiftBrowser/Source/ViewModel/DownloadCoordinator.swift: Added.
(DownloadCoordinator.downloads):
(DownloadCoordinator.didReceiveDownloadEvent(_:)):
(DownloadCoordinator.destination(forDownload:response:suggestedFilename:)):
(DownloadItem.url):
(DownloadItem.finished):
(DownloadItem.cancel):
(DownloadItem.hash(into:)):
* Tools/SwiftBrowser/Source/ViewModel/Downloadable.swift: Renamed from Tools/SwiftBrowser/Source/AppStorageKeys.swift.
(Downloadable.progress):
(Downloadable.url):
(Downloadable.finished):
(Downloadable.cancel):
* Tools/SwiftBrowser/Source/ViewModel/NavigationDecider.swift: Copied from Source/WebKit/UIProcess/API/Swift/WebPage+FrameInfo.swift.
(NavigationDecider.decidePolicy(for:preferences:)):
(NavigationDecider.decidePolicy(for:)):
* Tools/SwiftBrowser/Source/Views/BrowserView.swift: Renamed from Tools/SwiftBrowser/Source/BrowserView.swift.
(BrowserView.url):
(BrowserView.body):
(url):
* Tools/SwiftBrowser/Source/Views/ContentView.swift: Renamed from Tools/SwiftBrowser/Source/ContentView.swift.
(body):
(LabelConfiguration.body):
(PrincipalToolbarGroup.body):
(DialogActionsView.body):
(DialogMessageView.body):
(ContentView.url):
(ContentView.body):
(url):
* Tools/SwiftBrowser/Source/Views/DownloadsView.swift: Added.
(DownloadView.formattedSize):
(DownloadView.body):
(DownloadsList.body):
(Download.cancel):
* Tools/SwiftBrowser/Source/Views/SettingsView.swift: Renamed from Tools/SwiftBrowser/Source/SettingsView.swift.
(PermissionDecisionView.permissionDecision):
(PermissionDecisionView.body):
(GeneralSettingsView.body):
(SettingsView.body):
* Tools/SwiftBrowser/SwiftBrowser.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/288395@main">https://commits.webkit.org/288395@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/167e9c28a4764c65da7a3016c5c01bd786cedafd

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/83045 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/2686 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/37338 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/88154 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/34091 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/85137 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/2761 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/10607 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/88154 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/34091 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/86100 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/2021 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/75509 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/88154 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/1929 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/29711 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/33125 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/73111 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/30406 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/89520 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/10327 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/7463 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/73093 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/10556 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/71328 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/72317 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/16494 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/15232 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/1706 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12840 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/10280 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/15798 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/10152 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/13617 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/11921 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebKit/WebProcess/WebCoreSupport/WebContextMenuClient.cpp:
```diff
@@ -46,7 +46,6 @@ WTF_MAKE_TZONE_ALLOCATED_IMPL(WebContextMenuClient);
 void WebContextMenuClient::downloadURL(const URL&)
 {
     // This is handled in the UI process.
-    ASSERT_NOT_REACHED();
```
Please explain this in the commit message.

rr-codes committed:
[SwiftUI] Support download functionality
https://bugs.webkit.org/show_bug.cgi?id=285225
rdar://141109043

Reviewed by Aditya Keerthi.

In WebKit, there are several ways of initiating a download for a particular resource:

* Using the WebKit API directly to create a download using a specific `URLRequest`.

* Using the WebKit API directly to create a download using the data from a previously-cancelled download.

* A download can be created via web content before a request is sent by indicating a navigation action policy of `.download`.
This is typically used to download from a download attribute on an `<a>` tag.

* A download can be created via web content in the response after the HTTP headers have been received by indicating a navigation
response policy of `.download`. This is typically used to download a file with a 'Content-Disposition: attachment' HTTP header.

* A download can be created using one of the "Download linked file" default context menu items.

======

The new interface for creating and managing downloadable resources is described as follows:

* A download may be created or resumed by an API client using the respective functions on `WebPage`; these return an identifier
for the new download.

* In implementations of the navigation action and navigation response policy decision functions, clients can return a value of
`.download` to indicate the web content should be downloaded.

* A new protocol `DownloadCoordinator` is introduced.

* The `destination(forDownload:)` function of `DownloadCoordinator` is required to be implemented by clients so that WebKit can know
where to place the download. If this is not implemented, downloads are immediately cancelled. Downloads are also cancelled and an
exception raised if the file URL can not be created or already exists.

* The remaining functions of `DownloadCoordinator` are used to implement various policy decisions.

* WebPage now has an observable `downloads` property which is an async sequence of download events. Each download event has a
pecific kind to represent the download's progress, and an associated download value.

======

SwiftBrowser is updated to excercise this new functionality:

* Various types are refactored for better organization

* The app's sandbox is expanded to allow for access to the user's Downloads directory.

* Downloads now work as expected and download into the Downloads directory.

* A new Downloads window is added to view current downloads, with functionality to cancel ongoing downloads, and show finished downloads
in Finder.

=====

Misc:

* Remove a debug assertion that gets triggered when activating a context menu item that begins a download. While the comment above it
still holds true, the codepath will still be called regardless even though it doesn’t actually do anything, which is expected.

* Source/WebKit/UIProcess/API/Swift/DownloadCoordinator.swift: Added.
(DownloadCoordinator.destination(forDownload:response:suggestedFilename:)):
(DownloadCoordinator.authenticationChallengeDisposition(forDownload:challenge:URLCredential:)):
(DownloadCoordinator.httpRedirectionPolicy(forDownload:response:newRequest:)):
(DownloadCoordinator.placeholderPolicy(forDownload:)):
* Source/WebKit/UIProcess/API/Swift/WKDownloadDelegateAdapter.swift: Added.
(WKDownloadDelegateAdapter.owner):
(WKDownloadDelegateAdapter.downloadDidFinish(_:)):
(WKDownloadDelegateAdapter.download(_:didReceivePlaceholderURL:)):
(WKDownloadDelegateAdapter.download(_:didFailWithError:resumeData:)):
(WKDownloadDelegateAdapter.download(_:didReceiveFinalURL:)):
(WKDownloadDelegateAdapter.download(_:decideDestinationUsing:suggestedFilename:)):
(WKDownloadDelegateAdapter.download(_:willPerformHTTPRedirection:newRequest:)):
(WKDownloadDelegateAdapter.download(_:respondTo:URLCredential:)):
(WKDownloadDelegateAdapter.placeholderPolicy(forDownload:URL:)):
* Source/WebKit/UIProcess/API/Swift/WKNavigationDelegateAdapter.swift:
(WKNavigationDelegateAdapter.yieldDownloadProgress(_:download:)):
(WKNavigationDelegateAdapter.webView(_:navigationAction:didBecome:)):
(WKNavigationDelegateAdapter.webView(_:navigationResponse:didBecome:)):
* Source/WebKit/UIProcess/API/Swift/WebPage+Download.swift: Added.
(originalRequest):
(isUserInitiated):
(originatingFrame):
(progress):
(id):
(cancel):
(Downloads.makeAsyncIterator):
(Iterator.source):
(Iterator.next):
* Source/WebKit/UIProcess/API/Swift/WebPage+FrameInfo.swift:
(FrameInfo.wrapped):
* Source/WebKit/UIProcess/API/Swift/WebPage.swift:
(startDownload(using:)):
(resumeDownload(fromResumeData:)):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/SwiftBrowser/Resources/SwiftBrowser.entitlements:
* Tools/SwiftBrowser/Source/DownloadsView.swift: Added.
(DownloadsView.body):
* Tools/SwiftBrowser/Source/Extensions/URLResponse+Extras.swift: Copied from Source/WebKit/UIProcess/API/Swift/WebPage+FrameInfo.swift.
(URLResponse.hasAttachment):
* Tools/SwiftBrowser/Source/SwiftBrowser.swift:
(SwiftBrowserApp.body):
* Tools/SwiftBrowser/Source/ViewModel/AppStorageKeys.swift: Copied from Tools/SwiftBrowser/Source/AppStorageKeys.swift.
* Tools/SwiftBrowser/Source/ViewModel/BrowserViewModel.swift: Renamed from Tools/SwiftBrowser/Source/BrowserViewModel.swift.
(PDF.transferRepresentation):
(BrowserViewModel.decideSensorAuthorization(_:frame:origin:)):
(BrowserViewModel.displayedURL):
(BrowserViewModel.exportedPDF):
(BrowserViewModel.currentDialog):
(BrowserViewModel.currentFilePicker):
(BrowserViewModel.openURL(_:)):
(BrowserViewModel.didReceiveNavigationEvent(_:)):
(BrowserViewModel.navigateToSubmittedURL):
(BrowserViewModel.exportAsPDF):
(BrowserViewModel.didExportPDF(_:any:)):
(BrowserViewModel.didImportFiles(_:any:)):
(BrowserViewModel.setCameraCaptureState(_:)):
(BrowserViewModel.setMicrophoneCaptureState(_:)):
* Tools/SwiftBrowser/Source/ViewModel/DialogPresenter.swift: Added.
(hash(into:)):
(owner):
(handleJavaScriptAlert(_:initiatedBy:)):
(handleJavaScriptConfirm(_:initiatedBy:)):
(handleJavaScriptPrompt(_:defaultText:initiatedBy:)):
(handleFileInputPrompt(_:initiatedBy:)):
* Tools/SwiftBrowser/Source/ViewModel/DownloadCoordinator.swift: Added.
(DownloadCoordinator.downloads):
(DownloadCoordinator.didReceiveDownloadEvent(_:)):
(DownloadCoordinator.destination(forDownload:response:suggestedFilename:)):
(DownloadItem.url):
(DownloadItem.finished):
(DownloadItem.cancel):
(DownloadItem.hash(into:)):
* Tools/SwiftBrowser/Source/ViewModel/Downloadable.swift: Renamed from Tools/SwiftBrowser/Source/AppStorageKeys.swift.
(Downloadable.progress):
(Downloadable.url):
(Downloadable.finished):
(Downloadable.cancel):
* Tools/SwiftBrowser/Source/ViewModel/NavigationDecider.swift: Copied from Source/WebKit/UIProcess/API/Swift/WebPage+FrameInfo.swift.
(NavigationDecider.decidePolicy(for:preferences:)):
(NavigationDecider.decidePolicy(for:)):
* Tools/SwiftBrowser/Source/Views/BrowserView.swift: Renamed from Tools/SwiftBrowser/Source/BrowserView.swift.
(BrowserView.url):
(BrowserView.body):
(url):
* Tools/SwiftBrowser/Source/Views/ContentView.swift: Renamed from Tools/SwiftBrowser/Source/ContentView.swift.
(body):
(LabelConfiguration.body):
(PrincipalToolbarGroup.body):
(DialogActionsView.body):
(DialogMessageView.body):
(ContentView.url):
(ContentView.body):
(url):
* Tools/SwiftBrowser/Source/Views/DownloadsView.swift: Added.
(DownloadView.formattedSize):
(DownloadView.body):
(DownloadsList.body):
(Download.cancel):
* Tools/SwiftBrowser/Source/Views/SettingsView.swift: Renamed from Tools/SwiftBrowser/Source/SettingsView.swift.
(PermissionDecisionView.permissionDecision):
(PermissionDecisionView.body):
(GeneralSettingsView.body):
(SettingsView.body):
* Tools/SwiftBrowser/SwiftBrowser.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/288395@main
